### ISSUE TITLE
Allow `null` in registered constructor

### ISSUE THREAD

**Issue by rubenrubiob** [CuyZ/Valinor#245]
Hello,

We recently started using this library, and encountered the following issue.

Suppose we have a value object that accept string or null as its value:

```php
final class NullableValueObject
{
    private function __construct(private readonly ?string $value)
    {
    }

    public static function create(?string $value): self
    {
        return new self($value);
    }
}
```

And that with have some other object that has an attribute of that Value Object within:

```php
final class SomeDTO
{
    public function __construct(
        public readonly NullableValueObject $nullableValueObject
    ) {
    }
}
```

When we try to map the object with a null value, the library gives us an error:

```php
$mapper = (new \CuyZ\Valinor\MapperBuilder())
            ->registerConstructor([NullableValueObject::class, 'create'])
            ->flexible()
            ->mapper();

$source = [
    'nullableValueObject' => null,
];

$mapper->map(SomeDTO::class, Source::array($source));
```

```
CuyZ\Valinor\Mapper\MappingError : Could not map type `SomeDTO`. An error occurred at path nullableValueObject: Cannot be empty and must be filled with a value matching type `string|null`.
```

Disabling flexible mode gives us the same error. However, when the value of source is not null, the mapping works fine.

We are using a workaround: we build the object within the constructor of the DTO. Nonetheless, we would like to know if it would be possible to allow passing null values to registered constructors.

Thank you!

**Comment by rubenrubiob** [comment]
I tested using the just released 0.17.0 version. It works perfectly when using the new `enableFlexibleCasting`.

For reference, the mapper should be configured like so:

```
$mapper = (new \CuyZ\Valinor\MapperBuilder())
            ->registerConstructor([NullableValueObject::class, 'create'])
            ->enableFlexibleCasting()
            ->mapper();

$source = [
    'nullableValueObject' => null,
];

$mapper->map(SomeDTO::class, Source::array($source));
```

Thank you!

**Comment by romm** [comment]
Hi @rubenrubiob, I'm glad the last release fixes your issue, however your use case is an actual bug; your example should work even without the flexible mode. 

I'll reopen the issue and wait to have some time to take a further look.

Thanks for the report anyway!

**Comment by rubenrubiob** [comment]
I am somehow glad that we found a bug, then 😅 

We are heavily using Valinor, so if we find other edge cases we will let you know!

**Comment by romm** [comment]
Hi @rubenrubiob, this issue should be fixed with #250. When it is released, I recommend you do not use `enableFlexibleCasting` unless you really need it for something else. 😉

**Comment by rubenrubiob** [comment]
Great, thank you! We will update as soon as there is a new version available :)
